### PR TITLE
STM32F1: Fix ADC conversion for 12bits precision

### PR DESCRIPTION
**Pull request by tpruvot**
Follow up of #22789

### CODE REVIEW

**Comment by thisiskeithb** [comment]
Fixes #22797

**Comment by Malderin** [comment]
> Fixes #22797

Problem solved!

File: Marlin/src/HAL/STM32F1/HAL.cpp [hunk 1]
```diff
@@ -437,7 +437,7 @@ void HAL_adc_start_conversion(const uint8_t adc_pin) {
       case POWER_MONITOR_VOLTAGE_PIN: pin_index = POWERMON_VOLTS; break;
     #endif
   }
-  HAL_adc_result = (HAL_adc_results[(int)pin_index] >> 2) & 0x3FF; // shift to get 10 bits only.
+  HAL_adc_result = (HAL_adc_results[(int)pin_index] & 0xFFF) >> (12 - HAL_ADC_RESOLUTION); // shift out unused bits
```
Comment (thinkyhead) [inline Marlin/src/HAL/STM32F1/HAL.cpp]: Similar to `HAL/LPC1768` which has this:
```cpp
HAL_adc_reading = FilteredADC::read(pin) >> (16 - HAL_ADC_RESOLUTION); // returns 16bit value, reduce to required bits
```

**Comment by thinkyhead** [comment]
Seems safe to drop the `... & 0xFFF`, but is there any reason that `HAL_adc_results[]` could have a value larger than `2^12 - 1` … other than some kind of error?

**Pull request merged**